### PR TITLE
Fix some bugs

### PR DESCRIPTION
**Pull request by rahulmutt**
As mentioned in #1, we have transitioned from `cpphs` to `hpp` for `eta` and we ran into many minor bugs as we were testing against many Hackage packages. This PR should not be merged, and the comments made are to serve as discussion for how to handle certain aspects of the implementation. A lot of the fixes I made are quick and dirty and work for now, but we would want to look at a long-term fixes that are lot more clean.

### CODE REVIEW

File: src/Hpp.hs [hunk 1]
```diff
@@ -111,7 +112,7 @@ emptyHppState = T.emptyHppState
 
 -- | Create a 'T.HppState' with the given 'C.Config' and 'T.Env'.
 initHppState :: C.Config -> T.Env -> T.HppState
-initHppState c e = T.HppState c 1 e
+initHppState c e = setL lineNum 1 . setL env e . setL config c $ emptyHppState
```
Comment (rahulmutt) [inline src/Hpp.hs:115]: Done to make things a bit more consistent. The `config` lens setter in particular changes the value of the `dir` as well.

File: src/Hpp/CmdLine.hs [hunk 2]
```diff
@@ -89,7 +89,7 @@ parseArgs cfg0 = go emptyEnv id cfg0 Nothing . concatMap breakEqs
           return . Left $ BadCommandLine "Multiple output files given"
 
 -- | Run Hpp with the given commandline arguments.
-runWithArgs :: [String] -> IO ()
+runWithArgs :: [String] -> IO (Maybe Error)
```
Comment (rahulmutt) [inline src/Hpp/CmdLine.hs:92]: We use this function directly, and being able to get the error was very useful. Probably a better way to do this is to add a new function with a new name.

File: src/Hpp/RunHpp.hs [hunk 3]
```diff
@@ -258,18 +264,23 @@ runHpp source sink m = runHppT m >>= go []
...
+            cfg    <- use config
+            curDir <- use dir
+            let ipaths = includePaths cfg
+            mFound <- liftIO $ searchForNextInclude curDir ipaths file
+            readAux (file:files) ln file k mFound
           WriteOutput output k -> sink output >> runHppT k >>= go files
 
         readAux _files ln file _ Nothing =
-          Left . (, IncludeDoesNotExist ln file) . curFileName <$> use config
+          Left . (, IncludeDoesNotExist ln (stripAngleBrackets file))
```
Comment (rahulmutt) [inline src/Hpp/RunHpp.hs:287]: The `stripAngleBrackets` is just for slightly better error messages.

File: src/Hpp/RunHpp.hs [hunk 4]
```diff
@@ -444,10 +455,10 @@ directive = lift (onElements (awaitJust "directive")) >>= aux
                         let tokStr = toChars (detokenize toks)
                         throwError $ UserError ln (tokStr++" ("++curFile++")")
           "warning" -> True <$ lift dropLine -- warnings not yet supported
-          -- t -> do toks <- lift takeLine
-          --         ln <- subtract 1 <$> use lineNum
-          --         throwError $ UnknownCommand ln (detokenize (Important t:toks))
-          _ -> return False -- Ignore unknown command
+          t -> do toks <- lift takeLine
```
Comment (rahulmutt) [inline src/Hpp/RunHpp.hs]: I think this strict handling of unknown command should be behind some flag since I think normal C preprocessors allow unknown commands by simply ignoring them. Practically, it helps us out when we get errors on this since we know when CPP was used inappropriately.

File: src/Hpp/RunHpp.hs [hunk 5]
```diff
@@ -571,16 +582,15 @@ macroExpansion :: (Monad m, HppCaps m)
 macroExpansion = do
   lift await >>= \case
     Nothing -> return Nothing
+    Just ln@(Important "#":rst) -> do
```
Comment (rahulmutt) [inline src/Hpp/RunHpp.hs:644]: I'm not sure that this change is follows the spec, but I tried out an example with a preprocessor directive with some space prepended to it and it failed, so I assume most implementations expect there to be no whitespace from the beginning of the line to the first `#` of the directive.

I encountered this when compling a module with an input like:
```
{-# LANGUAGE MagicHash
    #-}
```
The 2nd line is treated as a directive and becomes an unknown command. Maybe this change should only be enabled when the flag I mentioned earlier is turned on?

File: src/Hpp/StringSig.hs [hunk 6]
```diff
@@ -195,6 +196,12 @@ sdropWhile f s = case sbreak (boolJust . f) s of
                    Just (_, _, s') -> s'
 {-# INLINE sdropWhile #-}
 
+stripR :: ByteString -> ByteString
```
Comment (rahulmutt) [inline src/Hpp/StringSig.hs:208]: Turns out `BL.lines` leaves the `\r` at the end and this breaks the line splicing code on Windows. Hence, I added an extra function in the pipeline to strip of the terminating `\r` if it exists.

File: src/Hpp/Tokens.hs [hunk 7]
```diff
@@ -95,16 +96,26 @@ tokWords s =
...
-                  let esc' = if sall isOctDigit esc
-                             then Important (digitsFromBase 8 esc)
-                             else case esc of
-                                    'x' :. hs
-                                      | sall isHexDigit hs ->
-                                      Important (digitsFromBase 16 hs)
-                                    (escapeChar -> Just e) :. Nil -> Important e
-                                    _ -> Important ("'\\" <> snoc esc '\'')
-                  in  maybeImp pre ++ esc' : tokWords pos'
+                Just (_,esc,pos')
```
Comment (rahulmutt) [inline src/Hpp/Tokens.hs:99]: This fixes the issue I mentioned where character literals are mutated into their numeric representations.

File: src/Hpp/Tokens.hs [hunk 8]
```diff
@@ -95,16 +96,26 @@ tokWords s =
...
+                  -- Deals with the '\'' case
```
Comment (rahulmutt) [inline src/Hpp/Tokens.hs:100]: The old code also didn't deal with `'\''` nicely and it would get converted to `0'` because the octal digits case would get hit - `all` returns `True` with an empty list.

File: src/Hpp/Types.hs [hunk 9]
```diff
@@ -85,8 +86,10 @@ instance Functor f => Functor (FreeF f a) where
 
 -- | Dynamic state of the preprocessor engine.
 data HppState = HppState { hppConfig :: Config
+                         , hppCurDir :: FilePath
```
Comment (rahulmutt) [inline src/Hpp/Types.hs:89]: This field was added to store the directory of the input. This + the changes above where I add this an extra parameter to `includeCandidates` in the code that looks up include files fix a bug described below:

```
hpp src/Main.hs
```

src/Main.hs:
```
#include "Main2.hs"
```

src/Main2.hs:
```
module Main where

main :: IO ()
main = return ()
```

Without the change I made to the include code, the example I provided above would convert `src/Main.hs` to an empty file or throw an IncludeDoesNotExist error if error values are not ignored. 

The reason is that upon the `#include "Main2.hs"`, `includeCandidates` will search for `Main2.hs` relative to the current directory of `hpp` which is the directory that contains `src`! Hence it won't find `Main2.hs`. I hit this bug when compiling the `vault` package which uses this method to organize code.

File: src/Hpp/RunHpp.hs (hunk 5 above)
Comment (acowley) [reply src/Hpp/RunHpp.hs:644]: I don't feel strongly about this, but making it configurable does seem like the right thing to do.

**Comment by acowley** [comment]
These are excellent changes, thank you for being so thorough! I hope you will consider adding test cases that reflect your uses so that we might more aggressively fix the gross infelicities in the library. I'm most eager to remove the half-baked streaming parser stuff, but I need to add a test case involving a large file before doing so in order to keep an eye on performance. Perhaps just profiling the existing mcpp test suite would suffice since pulling in any system header immediately involves a lot of work for the preprocessor.

**Comment by rahulmutt** [comment]
If you can setup the test framework/structure, I'd be happy to add test cases. 

The preprocessing step (if it runs) is almost *always* dwarfed by the time it takes to typecheck + optimize + code generate so I'm not too concerned about performance, as long as its correct. Morever, the input size has pretty small upper bound since most people like keeping each of their modules small for readability purposes. The only issue we may ever encounter is when people use CPP + Happy/Alex and in general anything that generates code. I have a test case for that and I'll run it sometime within the next couple weeks.

Since you seem OK with many of the changes, I'll create a new branch on my fork and submit a "clean" version of this PR.

**Comment by acowley** [comment]
Hi @rahulmutt, it's been a while so I'd like to know if I should close this. Several of these commits introduce failures with the validation suite we run against, so I'm not sure how many of the changes I can merge, but I'm trying to include as many as I can before changing `master` too much. I assume you've moved on, but would appreciate a confirmation of that.

**Comment by rahulmutt** [comment]
@acowley To run the validation suite, it's just `stack test` right?

I noticed that there are conflicts so I can rebase and fix them and see if I can fix the validation suite.

I would like to get these changes merged and remove the git dependency we're using right now.

**Comment by acowley** [comment]
You can look at the `.travis.yml` to see how it works, but it's a matter of `cd tests && ./mcpp-validation.sh`. This runs part of the mcpp (a C preprocessor) validation suite using `hpp` as the preprocessor. It does quite a nice job of exercising corner cases. My intention with it has always been that we'd pass its validation suite, then deviate from the needs of C with options to better serve Haskell.

**Comment by rahulmutt** [comment]
@acowley It looks like I won't have time to fix the validation suite - I did run it locally and there were too many problems, mainly with regards to the extra newlines we added in order to fix the placement of type error messages. I'm closing the PR and I'll submit a new one that is rebased against the current master once I find some time to take a close look.
